algorand/go-algorand
# ledger: convert FC unmarshalled nil value to empty byte slice on DB write KVs

## Summary



This is an improvement for DB on write, though it does not have effect on current sqlite3, it might matter if we take new DB backend (some Key-value based DB).

Since `encoded.KVRecordV6` is `codec:,omitempty,omitemptyarray`, when we have an instance of `encoded.KVRecordV6`, WLOG we talk about empty box, then an empty byte slice `[]byte{}` would be unmarshalled to `nil`, and written to DB.

The current status is, writing to DB with a `nil` value tied to a key doesn't really matter, for sqlite3 takes `nil` and `[]byte{}` as same empty BLOB. But for the sake of consistency, we should convert the value to `[]byte{}`.

Also, when we recheck with the logic of a round-by-round catchup, a box deletion would incur `kvmod.val = nil`, and `accountsNewRoundImpl` in `acctdelta.go` would treat this as a KV pair deletion in DB.

Thus, I think it would be good to convert a `nil` value unmarshal result to an empty byte slice `[]byte{}`.

## Test Plan



Consider the following test:
1. bring up a tracker DB instance of version 9
2. jam a bunch of keys with nil values into table `kvstore`
3. ensure that the DB read for nil values gets all the keys
4. start tracker DB migration from version 9 to version 10
5. ensure that the DB read for nil values gets 0 keys, i.e., all the nil values are converted to empty byte slices.
